README.md
# Logstash Plugin

[![Travis Build Status](https://travis-ci.com/logstash-plugins/logstash-input-udp.svg)](https://travis-ci.com/logstash-plugins/logstash-input-udp)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.

CHANGELOG.md
## 3.5.0
  - Added ECS v8 support as an alias to the ECS v1 implementation

## 3.4.1
  - [DOC] Fixed typo in code sample [#54](https://github.com/logstash-plugins/logstash-input-udp/pull/54)

## 3.4.0
  - Added ECS compatibility mode (`disabled` and `v1`) to rename ip source address in a ECS compliant name [#50](https://github.com/logstash-plugins/logstash-input-udp/pull/50)
  - Fixed integration tests for IPv6 downgrading Docker to version 2.4 [#51](https://github.com/logstash-plugins/logstash-input-udp/pull/51)

## 3.3.4
  - Fixed input workers exception handling and shutdown handling [#44](https://github.com/logstash-plugins/logstash-input-udp/pull/44)

## 3.3.3
  - Work around jruby/jruby#5148 by cloning messages on jruby 9k, therefore resizing the underlying byte buffer

## 3.3.2
  - Fix missing require for the ipaddr library.

## 3.3.1
  - Docs: Set the default_codec doc attribute.

## 3.3.0
  - Add metrics support for events, operations, connections and errors produced during execution. #34
  - Fix support for IPv6 #31

## 3.2.1
  - Code cleanup. See https://github.com/logstash-plugins/logstash-input-udp/pull/33

## 3.2.0
  - Clone codec per worker. See https://github.com/logstash-plugins/logstash-input-udp/pull/32

## 3.1.3
  - Update gemspec summary

## 3.1.2
  - Fix some documentation issues

## 3.1.0
  - Add "receive_buffer_bytes" config setting to optionally set socket receive buffer size

## 3.0.3
  - fix performance regression calling IO.select for every packet #21

## 3.0.2
  - Relax constraint on logstash-core-plugin-api to >= 1.60 <= 2.99

## 3.0.1
  - Republish all the gems under jruby.
## 3.0.0
  - Update the plugin to the version 2.0 of the plugin api, this change is required for Logstash 5.0 compatibility. See https://github.com/elastic/logstash/issues/5141
# 2.0.5
  - Depend on logstash-core-plugin-api instead of logstash-core, removing the need to mass update plugins on major releases of logstash
# 2.0.4
  - New dependency requirements for logstash-core for the 5.0 release
## 2.0.2
 - Adapt the test style to be able to run within the context of the LS
   core default plugins test (integration)

## 2.0.0
 - Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully, 
   instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
 - Dependency on logstash-core update to 2.0



spec/spec_helper.rb
# encoding: utf-8
require "logstash/devutils/rspec/spec_helper"
require 'logstash/inputs/udp'

# expose the udp socket so that we can assert, during
# a spec, that it is open and we can start sending data
class LogStash::Inputs::Udp
  attr_reader :udp
end

class UdpHelpers

  def input(plugin, size, &block)
    queue = Queue.new
    input_thread = Thread.new do
      plugin.run(queue)
    end
    # because the udp socket is created and bound during #run
    # we must ensure that it is open before sending data
    sleep 0.1 until (plugin.udp && !plugin.udp.closed?)
    block.call
    sleep 0.1 while queue.size != size
    result = size.times.inject([]) do |acc|
      acc << queue.pop
    end
    plugin.do_stop
    result
  end

end


spec/support/client.rb
# encoding: utf-8
require "socket"

module LogStash::Inputs::Test

  class UDPClient
    attr_reader :host, :port, :socket

    def initialize(port)
      @port = port
      @host = "0.0.0.0"
      @socket = UDPSocket.new
      socket.connect(host, port)
    end

    def send(msg)
      begin
        socket.connect(host, port) if socket.closed?
        socket.send(msg, 0)
      rescue  => e
        puts("send exception, retrying", e.inspect)
        retry
      end
    end

    def close
      socket.close unless socket.closed?
    end
  end
end


spec/inputs/udp_spec.rb
# encoding: utf-8
require_relative "../spec_helper"
require "logstash/devutils/rspec/shared_examples"
require_relative "../support/client"
require "logstash/codecs/base"


class LogStash::Codecs::Crash < LogStash::Codecs::Base
  config_name "crash"

  def decode(data)
    raise("decode crash") if data == "crash"
    yield LogStash::Event.new({"message" => data })
  end
end

describe LogStash::Inputs::Udp do
  before do
    srand(RSpec.configuration.seed)
  end

  let!(:helper) { UdpHelpers.new }
  let(:client) { LogStash::Inputs::Test::UDPClient.new(port) }
  let(:port) { rand(1024..65535) }
  let(:host) { "0.0.0.0" }
  let(:config) { { "port" => port, "host" => host } }
  subject { LogStash::Plugin.lookup("input","udp").new(config) }

  after :each do
    subject.close rescue nil
  end

  describe "register" do
    it "should register without errors" do
      expect { subject.register }.to_not raise_error
    end
  end

  describe "receive" do
  	shared_examples "receiving" do
      before(:each) do
        subject.register
      end

      let(:nevents) { 10 }

      let(:events) do
        helper.input(subject, nevents) do
          nevents.times do |i|
            client.send("msg #{i}")
          end
        end
      end

      it "should receive events been generated" do
        expect(events.size).to be(nevents)
        messages = events.map { |event| event.get("message")}
        messages.each do |message|
          expect(message).to match(/msg \d+/)
        end
      end
    end

    context "ipv4" do
      let(:client) { LogStash::Inputs::Test::UDPClient.new(port, "127.0.0.1") }
      include_examples "receiving"
    end
    context "ipv6" do
      let(:host) { "::1" }
      let(:client) { LogStash::Inputs::Test::UDPClient.new(port, "::1") }
      include_examples "receiving"
    end
  end

  shared_examples "respects ECS compatibility setting host metadata" do |ecs_compatibility, field_name|
    context "when ecs_compatibility is `#{ecs_compatibility}`" do
      let(:config) { { "port" => port, "workers" => 1, "ecs_compatibility" => ecs_compatibility} }
      let(:localhost) { "127.0.0.1" }
      let(:client) { LogStash::Inputs::Test::UDPClient.new(port, localhost) }

      let(:events) do
        helper.input(subject, 1) do
          client.send("line1")
        end
      end

      before(:each) do
        subject.register
      end

      it "produces event with source_ip_fieldname as '#{field_name}'" do
        expect(events.size).to be(1)
        message = events.last
        expect(message.get(field_name)).to eq(client.host)
      end
    end
  end

  it_behaves_like "respects ECS compatibility setting host metadata", :disabled, "host"
  it_behaves_like "respects ECS compatibility setting host metadata", :v1, "[host][ip]"
  it_behaves_like "respects ECS compatibility setting host metadata", :v8, "[host][ip]"



  describe "uses custom hostname field when ECS is enabled" do
    let(:config) { { "port" => port, "workers" => 1, "ecs_compatibility" => :v1, "source_ip_fieldname" => "custom_host_field"} }
    let(:localhost) { "127.0.0.1" }
    let(:client) { LogStash::Inputs::Test::UDPClient.new(port, localhost) }

    let(:events) do
      helper.input(subject, 1) do
        client.send("line1")
      end
    end

    before(:each) do
      subject.register
    end

    it "should receive event with user defined source_ip_fieldname" do
      expect(events.size).to be(1)
      message = events.last
      expect(message.get(config['source_ip_fieldname'])).to eq(client.host)
    end
  end

  describe "multiple lines per datagram using line codec" do
    # 3 workers for 3 datagrams send below
    let(:config) { { "port" => port, "workers" => 3, "codec" => "line" } }

    let(:events) do
      helper.input(subject, 8) do
        client.send("line1\nline2")
        client.send("line3\nline4")
        client.send("line5\nline6\nline7\nline8")
      end
    end

    before(:each) do
      subject.register
    end

    it "should receive events been generated" do
      expect(events.size).to be(8)
      messages = events.map { |event| event.get("message") }.sort # important to sort here because order is unpredictable
      messages.each_index {|i| expect(messages[i]).to match("line#{i + 1}")}
    end
  end

  it_behaves_like "an interruptible input plugin" do
    # see https://github.com/elastic/logstash-devutils/blob/9c4a1fbf2b0c4547e428c5a40ed84f60aad17f97/lib/logstash/devutils/rspec/shared_examples.rb
  end

  describe "worker should ignore codec exception" do
    # see custom "crash" codec above which raises upon decoding the "crash" straing payload
    let(:config) { { "port" => port, "workers" => 1, "codec" => "crash" } }

    let(:events) do
      helper.input(subject, 3) do
        client.send("crash")
        client.send("foo")
        client.send("bar")
        client.send("crash")
        client.send("baz")
      end
    end

    before(:each) do
      subject.register
    end

    it "should receive events been generated" do
      expect(events.size).to be(3)
      expect(events[0].get("message")).to match("foo")
      expect(events[1].get("message")).to match("bar")
      expect(events[2].get("message")).to match("baz")
    end
  end
end


lib/logstash/inputs/udp.rb
# encoding: utf-8
require "date"
require "logstash/inputs/base"
require "logstash/namespace"
require "socket"
require "stud/interval"
require "ipaddr"
require "logstash/plugin_mixins/ecs_compatibility_support"

# Read messages as events over the network via udp. The only required
# configuration item is `port`, which specifies the udp port logstash
# will listen on for event streams.
#
class LogStash::Inputs::Udp < LogStash::Inputs::Base
  # adds ecs_compatibility config which could be :disabled or :v1
  include LogStash::PluginMixins::ECSCompatibilitySupport(:disabled, :v1, :v8 => :v1)

  config_name "udp"

  default :codec, "plain"

  # The address which logstash will listen on.
  config :host, :validate => :string, :default => "0.0.0.0"

  # The port which logstash will listen on. Remember that ports less
  # than 1024 (privileged ports) may require root or elevated privileges to use.
  config :port, :validate => :number, :required => true

  # The maximum packet size to read from the network
  config :buffer_size, :validate => :number, :default => 65536

  # The socket receive buffer size in bytes.
  # If option is not set, the operating system default is used.
  # The operating system will use the max allowed value if receive_buffer_bytes is larger than allowed.
  # Consult your operating system documentation if you need to increase this max allowed value.
  config :receive_buffer_bytes, :validate => :number

  # Number of threads processing packets
  config :workers, :validate => :number, :default => 2

  # This is the number of unprocessed UDP packets you can hold in memory
  # before packets will start dropping.
  config :queue_size, :validate => :number, :default => 2000

  # The name of the field where the source IP address will be stored
  config :source_ip_fieldname, :validate => :string

  def initialize(params)
    super
    BasicSocket.do_not_reverse_lookup = true
  end

  def register
    @udp = nil
    @metric_errors = metric.namespace(:errors)
    if source_ip_fieldname.nil?
      # define ecs name mapping
      @field_source_ip = ecs_select[disabled: "host", v1: "[host][ip]"]
    else
      @field_source_ip = source_ip_fieldname
      if (ecs_compatibility != :disabled)
        @logger.warn("'source_ip_fieldname' is user customized, please check is has an ECS compatible name ")
      end
    end
  end # def register

  def run(output_queue)
    @output_queue = output_queue

    @input_to_worker = SizedQueue.new(@queue_size)
    metric.gauge(:queue_size, @queue_size)
    metric.gauge(:workers, @workers)

    @input_workers = (1..@workers).to_a.map do |i|
      @logger.debug("Starting UDP worker thread", :worker => i)
      Thread.new(i, @codec.clone) { |i, codec| inputworker(i, codec) }
    end

    begin
      # udp server
      udp_listener(output_queue)
    rescue => e
      @logger.error("UDP listener died", :exception => e, :backtrace => e.backtrace)
      @metric_errors.increment(:listener)
      Stud.stoppable_sleep(5) { stop? }
      retry unless stop?
    ensure
      # signal workers to end
      @input_workers.size.times { @input_to_worker.push([:END, nil]) }
      @input_workers.each { |thread| thread.join }
    end
  end

  def close
    if @udp && !@udp.closed?
      @udp.close rescue ignore_close_and_log($!)
    end
  end

  def stop
    if @udp && !@udp.closed?
      @udp.close rescue ignore_close_and_log($!)
    end
  end

  private

  def udp_listener(output_queue)
    @logger.info("Starting UDP listener", :address => "#{@host}:#{@port}")

    if @udp && !@udp.closed?
      @udp.close
    end

    if IPAddr.new(@host).ipv6?
      @udp = UDPSocket.new(Socket::AF_INET6)
    elsif IPAddr.new(@host).ipv4?
      @udp = UDPSocket.new(Socket::AF_INET)
    end
    # set socket receive buffer size if configured
    if @receive_buffer_bytes
      @udp.setsockopt(Socket::SOL_SOCKET, Socket::SO_RCVBUF, @receive_buffer_bytes)
    end
    rcvbuf = @udp.getsockopt(Socket::SOL_SOCKET, Socket::SO_RCVBUF).unpack("i")[0]
    if @receive_buffer_bytes && rcvbuf != @receive_buffer_bytes
      @logger.warn("Unable to set receive_buffer_bytes to desired size. Requested #{@receive_buffer_bytes} but obtained #{rcvbuf} bytes.")
    end

    @udp.bind(@host, @port)
    @logger.info("UDP listener started", :address => "#{@host}:#{@port}", :receive_buffer_bytes => "#{rcvbuf}", :queue_size => "#{@queue_size}")


    while !stop?
      next if IO.select([@udp], [], [], 0.5).nil?
      # collect datagram messages and add to inputworker queue
      @queue_size.times do
        begin
          payload, client = @udp.recvfrom_nonblock(@buffer_size)
          break if payload.empty?
          push_data(payload, client)
        rescue IO::EAGAINWaitReadable
          break
        end
      end
    end
  ensure
    if @udp
      @udp.close_read rescue ignore_close_and_log($!)
      @udp.close_write rescue ignore_close_and_log($!)
    end
  end

  def inputworker(number, codec)
    LogStash::Util::set_thread_name("<udp.#{number}")

    while true
      # a worker should never terminate from an exception, only when it receives the :END symbol
      begin
        payload, client = @input_to_worker.pop
        break if payload == :END

        ip_address = client[3]

        codec.decode(payload) { |event| push_decoded_event(ip_address, event) }
        codec.flush { |event| push_decoded_event(ip_address, event) }
      rescue => e
        @logger.error("Exception in inputworker", "exception" => e, "backtrace" => e.backtrace)
        @metric_errors.increment(:worker)
      end
    end
  end

  # work around jruby/jruby#5148
  # For jruby 9k (ruby >= 2.x) we need to truncate the buffer
  # after reading from the socket otherwise each
  # message will use 64kb
  if RUBY_VERSION.match(/^2/)
    def push_data(payload, client)
      payload = payload.b.force_encoding(Encoding::UTF_8)
      @input_to_worker.push([payload, client])
    end
  else
    def push_data(payload, client)
      @input_to_worker.push([payload, client])
    end
  end

  def push_decoded_event(ip_address, event)
    decorate(event)
    event.set(@field_source_ip, ip_address) if event.get(@field_source_ip).nil?
    @output_queue.push(event)
    metric.increment(:events)
  end

  def ignore_close_and_log(e)
    @logger.debug("ignoring close exception", "exception" => e)
  end
end
